chiron/reporters.py

```diff
@@ -99,11 +151,122 @@ def get_property(self, name: str):
 
         """
         if name not in self.h5file:
-            log.debug(f"{name} not in HDF5 file")
+            log.warning(f"{name} not in HDF5 file")
             return None
+        elif name == "u_kn":
+            return np.transpose(
+                np.array(self.h5file[name]), (2, 1, 0)
+            )  # shape: n_states, n_replicas, n_iterations
+
         else:
             return np.array(self.h5file[name])
 
+
+from typing import Optional
+
+
+class MultistateReporter(_SimulationReporter):
+    _name = "multistate_reporter"
+    _default_properties = [
+        "positions",
+        "box_vectors",
+        "u_kn",
+        "state_index",
+        "time",
+    ]
+
+    def __init__(self, buffer_size: int = 1) -> None:
+        filename = MultistateReporter.get_name()
+        directory = BaseReporter.get_directory()
+        import os
+
+        os.makedirs(directory, exist_ok=True)
+        self.file_path = directory / f"{filename}.h5"
+
+        super().__init__(file_path=self.file_path, buffer_size=buffer_size)
+        self._properties_to_report = MultistateReporter._default_properties
+        self._file_handle = {}
+
+    @classmethod
+    def get_name(cls):
+        return cls._name
+
+    def _write_to_trajectory(self, positions: np.ndarray):
+        import mdtraj as md
+
+        nr_of_runs, n_of_replicas, n_of_atoms, _ = positions.shape
+
+        # append to xtc trajectory the new positions
+        for replica_id in range(n_of_replicas):
+            file_name = f"replica_{replica_id}"
+            if self._file_handle.get(file_name) is None:
+                self._file_handle[file_name] = md.formats.XTCTrajectoryFile(
+                    f"{self.workdir}/{file_name}.xtc", mode="w"
+                )
+
+            open_xtc_file = self._file_handle[file_name]
+            xyz = positions[:, replica_id, :, :]
+            open_xtc_file.write(
+                xyz,
+                #            time=iteration,
+                box=self.get_property("box_vectors"),
+            )
+
+
+class MCReporter(_SimulationReporter):
+    _name = "mc_reporter"
+
+    def __init__(self, buffer_size: int = 1) -> None:
+        filename = MCReporter.get_name()
+        directory = BaseReporter.get_directory()
+        import os
+
+        os.makedirs(directory, exist_ok=True)
+        self.file_path = directory / f"{filename}.h5"
+
+        super().__init__(file_path=self.file_path, buffer_size=buffer_size)
+
+    @classmethod
+    def get_name(cls):
+        return cls._name
+
+
+class LangevinDynamicsReporter(_SimulationReporter):
+    _name = "langevin_reporter"
+    _default_properties = ["trajectory", "box_vectors", "potential_energy", "time"]
```

Comment:
Non-blocking, but making note that we probably want to in the future add in the computed temperature.